source/material_model/melt_phipps_morgan.cc

```diff
@@ -0,0 +1,1363 @@
+          else
+            out.densities[q] = melt_molar_mass / melt_molar_volume;
+
+          if (melt_out != nullptr)
+            {
+              double melt_compressiblity = 0;
+              for (unsigned int i=0; i<n_endmembers; ++i)
+                if (endmember_states[i] == EndmemberState::melt && melt_molar_volume > 0)
+                  melt_compressiblity += (endmember_mole_fractions_in_composite[i] * endmembers.volumes[i]) / (melt_molar_volume * endmembers.bulk_moduli[i]);
+
+              if (melt_molar_volume > 0)
+                melt_out->fluid_densities[q] = melt_molar_mass / melt_molar_volume;
+              else
+                {
+                  // make sure we have a useful melt density even if there is no melt to avoid density jumps
+                  double mass = 0, volume = 0;
+                  for (unsigned int i=0; i<n_endmembers; ++i)
+                    if (endmember_states[i] == EndmemberState::melt)
+                      {
+                        mass += endmember_mole_fractions_per_phase[i] * molar_masses[i];
+                        volume += endmember_mole_fractions_per_phase[i] * endmembers.volumes[i];
+                      }
+                  melt_out->fluid_densities[q] = mass/volume;
+                }
+
+              // TODO: this does not take into account the volume change due to thermal expansion of melt
+              melt_out->fluid_density_gradients[q] = melt_out->fluid_densities[q] * melt_out->fluid_densities[q]
+                                                     * melt_compressiblity
+                                                     * this->get_gravity_model().gravity_vector(in.position[q]);
+            }
+
+
+          if (this->include_melt_transport())
+            {
+              // Calculate the melting rate
+              const unsigned int Fe_solid_idx = this->introspection().compositional_index_for_name("molar_Fe_in_solid");
+              const unsigned int Fe_melt_idx = this->introspection().compositional_index_for_name("molar_Fe_in_melt");
+              const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
+
+              // TODO: Alternatively, this could also be done by summing over endmember_mole_fractions_in_composite
+              const double old_porosity = in.composition[q][porosity_idx];
+              const double old_solid_composition = in.composition[q][Fe_solid_idx];
+              const double old_melt_composition = in.composition[q][Fe_melt_idx];
+              const double old_melt_molar_fraction = melt_molar_fraction;
+
+              // in this simple model, the bulk composition is just one number, namely
+              // the molar fraction of the combined iron endmembers
+              const double bulk_composition = old_melt_composition * melt_molar_fraction + old_solid_composition * solid_molar_fraction;
+              // Q! solid_composition is already defined above!
```

Comment:
The solid and melt compositions need to be updated if there is melt transport (because the melt and solid can react with each other at each time step)